DOCS-11518-mqtt3-first-version-kt (#1579)
mulesoft/docs-connectors: mqtt3/1.0.0/modules/ROOT/pages/mqtt3-connector-publish-receive.adoc

= Listen and Publish Messages with MQTT3 Connector

In these examples, you configure the Anypoint Connector for MQTT3 (MQTT3 Connector) *On New Message* (`<mqtt3:listener>`) source and *Publish* (`<mqtt3:publish>`) operation to listen and publish messages to a specific topic.

== Create a Mule App to Listen and Publish Messages

For the first example, you must start an MQTT broker and create a Mule app. To create the Mule app in the Studio XML editor, follow these steps:

. In the Studio project, navigate to the *Configuration XML* tab.
. Configure an `<mqtt3:config name="MQTT3_Config_1">` and an `<http:listener-config>` configuration element:

[source,xml,linenums]
----
	<http:listener-config name="HTTP_Listener_config">
		<http:listener-connection host="0.0.0.0" port="8081" />
	</http:listener-config>
	<mqtt3:config name="MQTT3_Config_1">
		<mqtt3:connection clientId="crowley123" username="mosquitto" password="mosquitto" url="tcp://0.0.0.0:1883" />
		<mqtt3:last-will-and-testament topic="topic/lwt" body="This is the LWT message" isRetained="true"/>
	</mqtt3:config>
----

[start=3]

> . Create a flow with a listener source `<mqtt3:listener>` that triggers an event for incoming messages on topics matching the filter `topic/quotes/#`, and add a `<logger>` component that logs the received message::

Reviewer: ```suggestion
. Create a flow with a listener source `<mqtt3:listener>` that triggers an event for incoming messages on topics matching the filter `topic/quotes/#`, and add a `<logger>` component that logs the received message:
```

[source,xml,linenums]
----
	<flow name="listener-flow">
		<mqtt3:listener config-ref="MQTT3_Config_1" doc:name="On New Message 1">
			<reconnect />
			<mqtt3:topics >
				<mqtt3:topic topicFilter="topic/quotes/#" />
			</mqtt3:topics>
		</mqtt3:listener>
		<logger level="INFO" message="Received message '#[payload]' with at topic #[attributes.topic] with qos #[attributes.qos]"/>
	</flow>
----

[start=4]

> . Create a flow with a publish operation `<mqtt3:publish>` that publishes messages to the topic `topic/quotes/shakespeare` and triggers an event in the `listener-flow`:

Reviewer: Should this be *Publish* operation?

[source,xml,linenums]
----
	<flow name="publish-message-flow">
		<http:listener config-ref="HTTP_Listener_config" path="publish/message"/>
		<mqtt3:publish config-ref="MQTT3_Config_1" topic="topic/quotes/shakespeare">
			<mqtt3:message ><![CDATA[#["Uneasy lies the head that wears a crown"]]]></mqtt3:message>
		</mqtt3:publish>
		<set-payload value="SUCCESSFULLY PUBLISHED MESSAGE!" doc:name="Set Payload"/>
	</flow>
----

[start=5]
. To publish the message, send a request to the HTTP endpoint using the following curl command: +
 `curl -v --request GET localhost:8081/publish/message`.

== Create a Mule App and Subscribe to a LWT Message Topic

> In the following example, you use the previous listener flow and create a second listener flow subscribed to the same LWT message topic of the first flow. So if the Mule app crashes due to some reason, and the first listener flow gets disconnected, the MQTT broker sends the LWT message to the configured topic and retains the message. When the Mule app restarts, the second listener flow subscribed to the same topic, listens the retained LWT message and logs it.

Reviewer: ```suggestion
In the following example, you use the previous listener flow and create a second listener flow subscribed to the same LWT message topic as the first flow. If the Mule app crashes and the first listener flow gets disconnected, the MQTT broker sends the LWT message to the configured topic and retains the message. When the Mule app restarts, the second listener flow listens for the retained LWT message and logs it.
```

			<reconnect />
			<mqtt3:topics >
				<mqtt3:topic topicFilter="topic/quotes/#" />
			</mqtt3:topics>
		</mqtt3:listener>
		<logger level="INFO" message="Received message '#[payload]' with at topic #[attributes.topic] with qos #[attributes.qos]"/>
	</flow>
----

[start=4]
. Configure a new `<mqtt3:config name="MQTT3_Config_2">` configuration:

[source,xml,linenums]
----
<mqtt3:config name="MQTT3_Config_2">
	<mqtt3:connection clientId="azfell123" username="mosquitto" password="mosquitto" url="tcp://0.0.0.0:1884"/>
</mqtt3:config>
----

[start=5]
. Create a second listener flow by adding a listener source `<mqtt3:listener>` that subscribes to the topic `topic/lwt`, and add a `<logger>` component that logs the received message:

[source,xml,linenums]
----
<flow name="listener-lwt-flow">
	<mqtt3:listener doc:name="On New Message 2" config-ref="MQTT3_Config_2">
		<mqtt3:topics >
			<mqtt3:topic topicFilter="topic/lwt" />
		</mqtt3:topics>
	</mqtt3:listener>
	<logger level="INFO" message="Received message '#[payload]' with at topic #[attributes.topic] with qos #[attributes.qos]"/>
</flow>
----

[start=6]
. Save and run the Mule app.

> . Simulate the app crash by looking for the process ID of the Mule app and kill it in your terminal, for example `kill -9 <process-id>`. +

Reviewer: ```suggestion
. Simulate the app crash by looking for the process ID of the Mule app and kill it in your terminal by using a command like the following: `kill -9 <process-id>`.
+
```

Because the Mule app crashes, the first listener flow gets disconnected.

> . Restart the Mule app. +

Reviewer: ```suggestion
. Restart the Mule app. 
+
```
